New delegate API for event listener

Toxantron opened the pull request:
So this is what the the event listener API might look like. Feedback welcome.

Toxantron commented:
I will prepare possible usages of this new api on [a branch](https://github.com/Toxantron/angular-google-analytics/tree/delegate-api) to make sure i can push commits based on feedback to the master branch.

justinsa commented:
It is always good to step back for a moment when building a new feature and consider whether it is worth maintaining backwards compatibility. The ideal is that we move the library into an eventing model where extension functions are part of the documentation instead of there being multiple implementations embedded in the core library. A user can then extend through configuration and recipes.

To make the code cleaner, what if the `readFromRoute` behavior was moved into the documentation as a recipe and the `eventListener` is then redefined as a function bound to the scope of `that`, so:

```
eventListener.bind(that)();
```

It should probably be renamed to: `trackRouteFn` instead of listener, since it is not a listener.

Toxantron commented:
That would mean a little more refactoring. Especially since readFromRoute is not only part of the tracking but also changes the behavior of `getUrl()`. This would move the library towards a core/plugin design which has defined interfaces for replacing logic. Since I love doing stuff like this it would make the library more complex and my intentions where to maintain current features and only modify the internal structure so we can add different implementations of `trackRouteFn`(I like this name) as requirements popup. 

In addition to that I just wanted to extend this by opening the API to 'advanced users'. The first commit was only to demonstrate new API while maintaining full compatibility and unit test compliance.

Toxantron commented:
I added support for setting custom version of `trackRouteFn` to complete the original idea. This is what I had in mind as far as the refactoring goes. No when something like #145 pops up we can simple extend the block and we're good.

This has neither the flexibility nor the professionalism of your approach but it keeps the library simple and even though the code grows it remains maintainable.

justinsa commented on index.js:
```diff
@@ -191,6 +192,12 @@
+      
+      // Custom route tracking function
+      this.trackRouteFunction = function(trackingFunction) {
+        trackRouteFn = trackingFunction;
```
Check that `trackingFunction` is actually a function before assigning it. No need to throw an error if it is not, just don't set trackeRouteFn.

justinsa commented on index.js:
```diff
@@ -1081,20 +1088,31 @@
-              if (!$route.current || !$route.current.templateUrl || $route.current.doNotTrack) {
+        // This section adds different listeners based on the configuration
+        var eventListener;
+        if (trackRouteFn != null) {
```
Instead of checking for null, check that it is a function: `typeof trackRouteFn === 'function'`

justinsa commented on index.js:
```diff
@@ -1081,20 +1088,31 @@
+        }
+        
+        // Add listener if it was defined
+        if (eventListener != null) {
```
Instead of checking for null, check that it is a function: `typeof eventListener === 'function'`

Toxantron commented:
Ok, so I replaced `null`with `typeof === 'function'` and added the sample for hybrid mobile support from #145. 

Right now I am trying to think of a nice API to pass `_that` to `trackRouteFn`. This way #155 could be resolved by using the new delegate API like:

```js
AnalyticsProvider.trackRouteFunction(function(_that) {
// Some scope call
_that.trackPage();
});

However this won't work because the whole idea was to provide a function that matches the event signature. Any ideas?

Toxantron commented:
I think I have an idea how to pass in the interal object for *expert users*.

```js
if (typeof trackRouteFn === 'function') {
  // Redirct event to custom function, passing in protected object
  eventListener = function () {
    var args = Array.prototype.slice.call(arguments);
    var newArgs = [ that ].concat(args);
    trackRouteFn.apply(null, newArgs);
  }
}
```

justinsa commented on index.js:
```diff
@@ -1081,20 +1090,38 @@
-              // and those explicitly marked as 'do not track'
-              if (!$route.current || !$route.current.templateUrl || $route.current.doNotTrack) {
+        // This section adds different listeners based on the configuration
+        var eventListener;
```
This should be set to null, for good practice.

justinsa commented on index.js:
```diff
@@ -1081,20 +1090,38 @@
+        var eventListener;
+        if (typeof trackRouteFn === 'function') {
+          eventListener = trackRouteFn;
+        } else if (trackRoutes && hybridMobileSupport) {
```
The following else-if sections are all sub-branches off of `trackRoutes` being true. Do that check and then do an inner if-else block to handle the variants. e.g.,

```javascript
} else if (trackRoutes) {
   if (hybridMobileSupport) {
      ...
   } else if (!readFromRoute) {
      ...
   } else { // readFromRoute
      ...
   }
}
```

That will be more readable as to what is going on in the logic.

justinsa commented on index.js:
```diff
@@ -1081,20 +1090,38 @@
+        if (typeof trackRouteFn === 'function') {
+          eventListener = trackRouteFn;
+        } else if (trackRoutes && hybridMobileSupport) {
+          // For mobile devices we need to track a sreen rather than a page
```
screen is misspelled in this comment.

justinsa commented:
@Toxantron why not use binding so the `this` in the function is `that`.

```javascript
if (typeof trackRouteFn === 'function') {
  // Redirect event to custom function, passing in protected object
  eventListener = function () {
    trackRouteFn.apply(that, arguments);
  }
}
```

Toxantron commented:
Good point. I thought of binding like `trackRouteFn.bind(that)` which doesn't work since angular will assign a new `this` when adding the listener.
However your proposal is shorter and easier to explain to developers.

justinsa commented on index.js:
```diff
@@ -1081,20 +1090,43 @@
               // and those explicitly marked as 'do not track'
               if (!$route.current || !$route.current.templateUrl || $route.current.doNotTrack) {
-                return;
+                  return;
```
This got 4 space indentation accidentally.

Toxantron replied:
Man, indentation is a pain in the ass. Will fix ASAP.

justinsa commented:
@Toxantron this is looking solid. Are there tests for everything? Can you add one for the "screeName" being sent in hybridMobileSupport case? Last ask, I promise.

Toxantron commented:
Thanks, but I don't think it is ready to merge just yet. We allready have unit tests for most of it like tracking and route reading. I will add some for the `customTrackFn` feature and maybe test a use case similar to #155. For #145 we will need to add `appName` to the configuration as pointed out in #161.

Toxantron commented:
So I fixed indentation and added unit tests. Current state would close #155.

@justinsa where should we add the `appName`? The account object?

Toxantron commented:
I had a look at the `account` object and it didn't fit. I added it to the config and adapted tracking function and unit test. 

This fixes #145 and #161.

justinsa commented on index.js:
```diff
@@ -14,6 +14,7 @@
     .provider('Analytics', function () {
       var accounts,
           analyticsJS = true,
+          appName,
```
Change this to hybridMobileAppName. That makes it clearly related to hybridMobileSupport and not a generic configuration value.

justinsa replied:
Also, please expose this value in the configuration property of the return object. All variables are exposed there (near the end of the file). This variable also needs to be listed in the README.md as available under the configuration property.

Toxantron replied:
Sure.

justinsa commented:
@Toxantron Do you know if it is legal for the appName property to be undefined in the call to GA?

Toxantron commented:
I see what you are getting at. I think from documentation it stated that
appName is mandatory. I will log a warning and add a unit test.
On Jun 17, 2016 08:10, "Justin Saunders" <notifications@github.com> wrote:

> @Toxantron <https://github.com/Toxantron> Do you know if it is legal for
> the appName property to be undefined in the call to GA?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/revolunet/angular-google-analytics/pull/149#issuecomment-226688273>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AOPV-Nqf3ukvRU-O1uDtTm3rp2VZv2Bpks5qMjpHgaJpZM4IDy3P>
> .
>

justinsa commented on index.js:
```diff
@@ -1091,27 +1103,52 @@
+          if (hybridMobileSupport) {
+            // For mobile devices we need to track a screen rather than a page
+            eventListener = function (event, state) {
+              that._send('screenview', {
```
Log the warning inside the eventListener function, if you weren't already thinking of doing that. Do you think it should still make the _send call?

Toxantron replied:
That's what I had in mind.
I haven't decided yet. I will setup a little test locally and see how
google analytics handles the missing field. I will also add a test to
ensure the values are present in Analytics.configuration.
Once it is done we should be good to go.

Toxantron commented:
Or maybe @josephlaw can answer the question.. He seems to have more experience with the mobile mode.

josephlaw commented:
My hybrid app temporary use below
```
   // GA with hybrid app using screenview(app) instead of pageview(web)
    // https://developers.google.com/analytics/devguides/collection/analyticsjs/screens#using_filters_for_app-only_or_web-only_views
    $rootScope.$on('$stateChangeSuccess', function(event, state) {
        Analytics.send('screenview', {
            screenName: state.name,
            appName: 'xxxxx',
            appId: 'X.34567.89',
            appVersion: '0.1',
            appInstallerId: 'apk',
        });
    });

```

Toxantron commented:
@josephlaw what would you expect the library to do when appName was not set? Can you tell how google analytics handles it?

Toxantron commented:
Wrong button.

josephlaw commented:
If appName not set then the GA do nothing.

Toxantron commented:
Ok, then we only log exception and not send the call at all. Agreed @justinsa?

justinsa commented:
We should allow for any arbitrary optional values to be defined as well:

e.g.,
```
            appId: 'X.34567.89',
            appVersion: '0.1',
            appInstallerId: 'apk',
```

Otherwise this isn't going to be very useful to @josephlaw since he will need to override the behavior anyways. Perhaps instead of mobileAppName the user can provide mobileAppProperties.

Toxantron commented:
I agree. And then simply copy properties into the call.

Toxantron commented:
@justinsa what do you think?

Toxantron commented on index.js:
```diff
@@ -1091,27 +1123,56 @@
+              }
+              // Create a copy for the tracking function to avoid manipulation during transmission
+              var args = angular.copy(mobileAppProperties);
+              args.screenName = state.name;
```
We may need to differentiate between `ngRoute` and `ui-router` here. The sample taken from #145 seems to use `ui-router` while our default is `ngRoute`.

Toxantron commented:
@justinsa how do we proceed here?

justinsa commented:
@toxantron This is why we should remove these altogether and instead have a cookbook available that examples specific configurations. Only hybrid mobile cares about this and that should be configured by the developer for their specific scenario. If we simplify all the configuration code we have added and move that to a cookbook then support one reasonable default per handler (like ngRoute), this library becomes better. We can also rev the major version so it is clear the change breaks the past.

Toxantron commented:
You mean moving just the different delegates or the configuration as a whole? Would you mind adding a small sample of what you have in mind?
